### PR TITLE
[Plugin] [0/9] Split plugin into separate files and add bundling option.

### PR DESCRIPTION
**Pull request by tjzel**
## Summary

More work on plugin refactorization. This is simply splitting the code into files without any changes to it and adding `esbuild` bundler to generate a single file.

## Test plan

Run `yarn jest` and build Example app.

### CODE REVIEW

File: plugin/package.json [hunk 1]
```diff
@@ -2,13 +2,15 @@
...
+    "typescript": "^4.1.3",
+    "esbuild": "^0.17.11"
   },
   "license": "MIT",
   "readmeFilename": "README.md",
   "scripts": {
     "prepare": "yarn generate",
-    "generate": "yarn tsc",
+    "generate": "yarn tsc && yarn bundle",
+    "bundle": "yarn esbuild src/plugin.js --bundle --external:@babel --external:convert-source-map --external:fs --platform=node --sourcemap=linked --outfile=index.js && rm src/*.js src/*.js.map",
```
Comment (tjzel) [inline plugin/package.json:13]: I am not sure if removing compiled source files is a fail-safe option.

File: plugin/package.json [hunk 2]
```diff
@@ -2,13 +2,15 @@
...
+    "typescript": "^4.1.3",
+    "esbuild": "^0.17.11"
   },
   "license": "MIT",
   "readmeFilename": "README.md",
   "scripts": {
     "prepare": "yarn generate",
-    "generate": "yarn tsc",
+    "generate": "yarn tsc && yarn bundle",
+    "bundle": "yarn esbuild src/plugin.js --bundle --external:@babel --external:convert-source-map --external:fs --log-level=warning --platform=node --sourcemap=linked --outfile=index.js && rm src/*.js src/*.js.map",
```
Comment (tomekzaw) [inline plugin/package.json:13]: Is it possible to specify these options in some kind of esbuild config file?

Comment (tomekzaw) [inline plugin/package.json:13]: Can we move `rm src/*.js src/*.js.map` to a separate `clean` script and call it inside `generate`?

File: plugin/src/commonFunctions.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  let hash2 = 52711;
+
+  while (i--) {
+    const char = str.charCodeAt(i);
+    hash1 = (hash1 * 33) ^ char;
+    hash2 = (hash2 * 33) ^ char;
+  }
+
+  return (hash1 >>> 0) * 4096 + (hash2 >>> 0);
+}
```
Comment (tomekzaw) [inline plugin/src/commonFunctions.ts:13]:
```suggestion
}

```

File: plugin/src/commonFunctions.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+  }
+
+  if (process.env.REANIMATED_PLUGIN_TESTS === 'jest') {
+    // We want to detect this, so we can disable source maps (because they break
+    // snapshot tests with jest).
+    return false;
+  }
+
+  return true;
+}
```
Comment (tomekzaw) [inline plugin/src/commonFunctions.ts:33]:
```suggestion
}

```

File: plugin/src/commonFunctions.ts [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+export { hash, isRelease, shouldGenerateSourceMap };
```
Comment (tomekzaw) [inline plugin/src/commonFunctions.ts:34]: Please use `export function` as we already do throughout our codebase 🙏

**Review by tomekzaw — CHANGES_REQUESTED** [review]

File: plugin/src/commonInterfaces.ts [hunk 6]
```diff
@@ -0,0 +1,15 @@
+import * as BabelCore from '@babel/core';
```
Comment (tomekzaw) [inline plugin/src/commonInterfaces.ts:1]: types.ts

File: plugin/src/commonObjects.ts [hunk 7]
```diff
@@ -0,0 +1,59 @@
+export const globals = new Set([
```
Comment (tomekzaw) [inline plugin/src/commonObjects.ts:1]: rename file plz

File: plugin/package.json [hunk 8]
```diff
@@ -2,13 +2,16 @@
...
+    "typescript": "^4.1.3",
+    "esbuild": "^0.17.11"
   },
   "license": "MIT",
   "readmeFilename": "README.md",
   "scripts": {
     "prepare": "yarn generate",
-    "generate": "yarn tsc",
+    "generate": "yarn tsc && yarn bundle && yarn clean",
+    "clean": "rm src/*.js src/*.js.map",
```
[the remark this answers is not in the document]
Comment (tjzel) [reply plugin/package.json:13]:
```
———————————lib?———————————
⠀⣞⢽⢪⢣⢣⢣⢫⡺⡵⣝⡮⣗⢷⢽⢽⢽⣮⡷⡽⣜⣜⢮⢺⣜⢷⢽⢝⡽⣝
⠸⡸⠜⠕⠕⠁⢁⢇⢏⢽⢺⣪⡳⡝⣎⣏⢯⢞⡿⣟⣷⣳⢯⡷⣽⢽⢯⣳⣫⠇
⠀⠀⢀⢀⢄⢬⢪⡪⡎⣆⡈⠚⠜⠕⠇⠗⠝⢕⢯⢫⣞⣯⣿⣻⡽⣏⢗⣗⠏⠀
⠀⠪⡪⡪⣪⢪⢺⢸⢢⢓⢆⢤⢀⠀⠀⠀⠀⠈⢊⢞⡾⣿⡯⣏⢮⠷⠁⠀⠀
⠀⠀⠀⠈⠊⠆⡃⠕⢕⢇⢇⢇⢇⢇⢏⢎⢎⢆⢄⠀⢑⣽⣿⢝⠲⠉⠀⠀⠀⠀
⠀⠀⠀⠀⠀⡿⠂⠠⠀⡇⢇⠕⢈⣀⠀⠁⠡⠣⡣⡫⣂⣿⠯⢪⠰⠂⠀⠀⠀⠀
⠀⠀⠀⠀⡦⡙⡂⢀⢤⢣⠣⡈⣾⡃⠠⠄⠀⡄⢱⣌⣶⢏⢊⠂⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢝⡲⣜⡮⡏⢎⢌⢂⠙⠢⠐⢀⢘⢵⣽⣿⡿⠁⠁⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠨⣺⡺⡕⡕⡱⡑⡆⡕⡅⡕⡜⡼⢽⡻⠏⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⣼⣳⣫⣾⣵⣗⡵⡱⡡⢣⢑⢕⢜⢕⡝⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⣴⣿⣾⣿⣿⣿⡿⡽⡑⢌⠪⡢⡣⣣⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⡟⡾⣿⢿⢿⢵⣽⣾⣼⣘⢸⢸⣞⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠁⠇⠡⠩⡫⢿⣝⡻⡮⣒⢽⠋⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
—————————————————————————————
```

File: plugin/tsconfig.json [hunk 9]
```diff
@@ -7,7 +7,9 @@
     "sourceMap": true,
     "strict": true,
     "resolveJsonModule": true,
-    "esModuleInterop": true
+    "esModuleInterop": true,
+    "outDir": "lib/",
+    "rootDir": "src"
```
Comment (tomekzaw) [inline plugin/tsconfig.json:12]:
```suggestion
    "rootDir": "src/"
```

**Review by tomekzaw — APPROVED** [review]
👏

**Pull request merged**